### PR TITLE
Fix direct memory leaks in deterministic-unit_test_hamiltonian_force

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

This fixes most direct leaks found in deterministic-unit_test_hamiltonian_force by clang address sanitizer.

I didn't fix the leak in src/QMCWaveFunctions/Jastrow/RadialJastrowBuilder.cpp:343 since @williamfgc already has a branch that changes the return type from a raw pointer to a unique_ptr. 

This is part of #3083.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- n/a. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- n/a. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by ye-luo** [comment]
Test this please

**Comment by ye-luo** [comment]
Test this please

**Comment by prckent** [comment]
Test this please

File: src/QMCHamiltonians/ECPComponentBuilder.cpp [hunk 1]
```diff
@@ -37,7 +37,7 @@ ECPComponentBuilder::ECPComponentBuilder(const std::string& aname, Communicate*
       Zeff(0),
       RcutMax(-1),
       Species(aname),
-      grid_global(0)
+      grid_global()
```
Comment (ye-luo) [inline src/QMCHamiltonians/ECPComponentBuilder.cpp:40]: Just remove.

File: src/QMCHamiltonians/ECPComponentBuilder.1.cpp [hunk 2]
```diff
@@ -154,12 +154,12 @@ void ECPComponentBuilder::buildLocal(xmlNodePtr cur)
   {
     if (grid_global == 0)
```
Comment (ye-luo) [inline src/QMCHamiltonians/ECPComponentBuilder.1.cpp:155]: nullptr instead of 0.

**Review by ye-luo — COMMENTED** [review]
Only minor revision needed. Please format all the files being touched by this PR

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**